### ISSUE TITLE
Invalidation failure

### ISSUE THREAD

**Issue by ferrao** [apancutt/deploy-aws-s3-cloudfront#6]
My deploys work just fine 50% of the times. The other 50%, i get:

`ERROR Invalidation failure (Your request contains one or more invalid invalidation paths.)`

And no invalidation occurs, so I have to invalidate manually. I could not find any correlation between anything that I do and the fails.

**Comment by ferrao** [comment]
What log are you referring to @apancutt ?

I always get a bunch of nice green messages about the deployment to S3 and then about the path invalidation and everything seems to go well.

But then sometimes it prints that error message right at the end and npm bails on error. The extra error message and the fact that no invalidation appears on the aws control panel is the only difference I can see between the cases it works well and the ones where it fails.

**Comment by ferrao** [comment]
This happens to me in different projects @apancutt, the only thing in common I can think of is that they all use webpack for bundling. I can provide you with the full webpack output the next time I deploy and it occurs, I just have something within a few days to a week.

**Comment by apancutt** [comment]
I'm referring to the log that you get when running this tool. It will detail the changes uploaded and what will be invalidated.

Please also include the exact command you execute (obfuscate anything sensitive first, though).

**Comment by ferrao** [comment]
Just had to do another deployment and stumbled across the issue again, here it goes:

```
> deploy-aws-s3-cloudfront --profile=xxxxx --bucket admissions.xxxxx-staging --source=./build/ --distribution YYYYYYYYYYYYYY

Fetching local objects in /Users/ferrao/Dev/Workspace/xxxxx/xxxxx-admissions/build/...
Fetching remote objects in admissions.xxxxx-staging...
Checking for modifications...
M 0.css
M 0.css.map
M 1.css
M 1.css.map
M 3.css
M 3.css.map
A commons.03be2cf23455edb0b5fb.chunk.js.LICENSE
A commons.03be2cf23455edb0b5fb.chunk.js
A commons.03be2cf23455edb0b5fb.chunk.js.map
M index.html
A main.da1aa8c4f14f1dbe84fe.chunk.js
A runtime~main.50dbf6b8fde4f1a82a57.bundle.js
A vendor.fa9032f9d90e1ebcbdc7.chunk.js
A vendor.fa9032f9d90e1ebcbdc7.chunk.js.map
A vendor.fa9032f9d90e1ebcbdc7.chunk.js.LICENSE
? Continue deployment? true
Uploading commons.03be2cf23455edb0b5fb.chunk.js.LICENSE (2.4 kB) as application/octet-stream to admissions.xxxxx-staging...
Uploading commons.03be2cf23455edb0b5fb.chunk.js (1.52 MB) as application/javascript to admissions.xxxxx-staging...
Uploading commons.03be2cf23455edb0b5fb.chunk.js.map (158 B) as application/json to admissions.xxxxx-staging...
Uploading main.da1aa8c4f14f1dbe84fe.chunk.js (587 kB) as application/javascript to admissions.xxxxx-staging...
Uploading runtime~main.50dbf6b8fde4f1a82a57.bundle.js (1.49 kB) as application/javascript to admissions.xxxxx-staging...
Uploading vendor.fa9032f9d90e1ebcbdc7.chunk.js (2.03 MB) as application/javascript to admissions.xxxxx-staging...
Uploading vendor.fa9032f9d90e1ebcbdc7.chunk.js.map (156 B) as application/json to admissions.xxxxx-staging...
Uploading vendor.fa9032f9d90e1ebcbdc7.chunk.js.LICENSE (4.8 kB) as application/octet-stream to admissions.xxxxx-staging...
Uploading 0.css (14.5 kB) as text/css to admissions.xxxxx-staging...
Uploading 0.css.map (18.4 kB) as application/json to admissions.xxxxx-staging...
Uploading 1.css (13 kB) as text/css to admissions.xxxxx-staging...
Uploading 1.css.map (17.5 kB) as application/json to admissions.xxxxx-staging...
Uploading 3.css (213 kB) as text/css to admissions.xxxxx-staging...
Uploading 3.css.map (223 kB) as application/json to admissions.xxxxx-staging...
Uploading index.html (3.21 kB) as text/html to admissions.xxxxx-staging...
Invalidating /commons.03be2cf23455edb0b5fb.chunk.js.map on CloudFront distribution YYYYYYYYYYYYYYY...
Invalidating /runtime~main.50dbf6b8fde4f1a82a57.bundle.js on CloudFront distribution YYYYYYYYYYYYYYY...
Invalidating /vendor.fa9032f9d90e1ebcbdc7.chunk.js.map on CloudFront distribution YYYYYYYYYYYYYYY...
Invalidating /commons.03be2cf23455edb0b5fb.chunk.js.LICENSE on CloudFront distribution YYYYYYYYYYYYYYY...
Invalidating /index.html on CloudFront distribution YYYYYYYYYYYYYYY...
Invalidating /vendor.fa9032f9d90e1ebcbdc7.chunk.js.LICENSE on CloudFront distribution YYYYYYYYYYYYYYY...
Invalidating /1.css on CloudFront distribution YYYYYYYYYYYYYYY...
Invalidating /0.css on CloudFront distribution YYYYYYYYYYYYYYY...
Invalidating /0.css.map on CloudFront distribution YYYYYYYYYYYYYYY...
Invalidating /1.css.map on CloudFront distribution YYYYYYYYYYYYYYY...
Invalidating /3.css on CloudFront distribution YYYYYYYYYYYYYYY...
Invalidating /3.css.map on CloudFront distribution YYYYYYYYYYYYYYY...
Invalidating /main.da1aa8c4f14f1dbe84fe.chunk.js on CloudFront distribution YYYYYYYYYYYYYYY...
Invalidating /commons.03be2cf23455edb0b5fb.chunk.js on CloudFront distribution YYYYYYYYYYYYYYY...
Invalidating /vendor.fa9032f9d90e1ebcbdc7.chunk.js on CloudFront distribution YYYYYYYYYYYYYYY...
ERROR Invalidation failure (Your request contains one or more invalid invalidation paths.)
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! xxxxx-webapp@0.0.1 deploy: `deploy-aws-s3-cloudfront --profile=xxxxx --bucket admissions.xxxxx-staging --source=./build/ --distribution YYYYYYYYYYYYYYY`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the xxxxx-webapp@0.0.1 deploy script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/ferrao/.npm/_logs/2019-11-10T12_50_11_713Z-debug.log
```

The relevant part of the npm log is:

```
10 silly lifecycle   'deploy-aws-s3-cloudfront --profile=xxxxx --bucket admissions.xxxxx-staging --source=./build/ --distribution YYYYYYYYYYYYYY' ]
11 silly lifecycle xxxxx-webapp@0.0.1~deploy: Returned: code: 1  signal: null
12 info lifecycle xxxxx-webapp@0.0.1~deploy: Failed to exec deploy script
13 verbose stack Error: xxxxx-webapp@0.0.1 deploy: `deploy-aws-s3-cloudfront --profile=xxxxx --bucket admissions.xxxxx-staging --source=./build/ --distribution YYYYYYYYYYYYYY`
13 verbose stack Exit status 1
13 verbose stack     at EventEmitter.<anonymous> (/Users/ferrao/.nvm/versions/node/v10.14.1/lib/node_modules/npm/node_modules/npm-lifecycle/index.js:332:16)
13 verbose stack     at EventEmitter.emit (events.js:182:13)
13 verbose stack     at ChildProcess.<anonymous> (/Users/ferrao/.nvm/versions/node/v10.14.1/lib/node_modules/npm/node_modules/npm-lifecycle/lib/spawn.js:55:14)
13 verbose stack     at ChildProcess.emit (events.js:182:13)
13 verbose stack     at maybeClose (internal/child_process.js:962:16)
13 verbose stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:251:5)
14 verbose pkgid xxxxx-webapp@0.0.1
15 verbose cwd /Users/ferrao/Dev/Workspace/academia-codigo/xxxxx/xxxxx-admissions
16 verbose Darwin 18.5.0
17 verbose argv "/Users/ferrao/.nvm/versions/node/v10.14.1/bin/node" "/Users/ferrao/.nvm/versions/node/v10.14.1/bin/npm" "run" "deploy"
18 verbose node v10.14.1
19 verbose npm  v6.11.3
20 error code ELIFECYCLE
21 error errno 1
22 error xxxxx-webapp@0.0.1 deploy: `deploy-aws-s3-cloudfront --profile=xxxxx --bucket admissions.xxxxx-staging --source=./build/ --distribution YYYYYYYYYYYYYY`
22 error Exit status 1
23 error Failed at the xxxxx-webapp@0.0.1 deploy script.
23 error This is probably not a problem with npm. There is likely additional logging output above.
24 verbose exit [ 1, true ]
````

**Comment by apancutt** [comment]
Thanks @ferrao - looks like a problem with the tilde (`~`) in one of your invalidation paths and AWS not accepting it. [Turns out](https://forums.aws.amazon.com/thread.jspa?messageID=219736) the paths must be [RFC1738](http://www.faqs.org/rfcs/rfc1738.html)-compliant and `encodeURIComponent()` is [missing a few characters](https://locutus.io/php/url/rawurlencode/), the tilde being one of them.

This has been fixed in 1.0.1. Please upgrade and check again.

**Comment by apancutt** [comment]
Fixed by 1.0.1

**Comment by ferrao** [comment]
Those file names with the tilde are generated by webpack, not something that I set explicitly, strange that nobody tripped on this. 

Anyway, thanks for the quick turnaround @apancutt !
